pkg/extensions/upgrade.go

```diff
@@ -45,113 +32,170 @@ func Upgrade(ctx context.Context, kcli client.Client, prev *ecv1beta1.Installati
-		reason := actionEd
-		message := ""
-
-		if finalErr != nil {
-			status = metav1.ConditionFalse
-			reason = action + "Failed"
-			message = helpers.CleanErrorMessage(finalErr)
-		}
-
-		if err := setCondition(ctx, kcli, in, conditionName(ext), status, reason, message); err != nil {
-			slog.Error("Failed to set condition status", "error", err)
-		}
-	}()
-
-	switch action {
-	case actionInstall:
+func handleExtensionInstall(ctx context.Context, kcli client.Client, hcli helm.Client, in *ecv1beta1.Installation, ext ecv1beta1.Chart) (finalErr error) {
+	return handleExtension(ctx, kcli, in, ext, actionInstall, func() error {
 		exists, err := hcli.ReleaseExists(ctx, ext.TargetNS, ext.Name)
 		if err != nil {
 			return errors.Wrap(err, "check if release exists")
 		}
 		if exists {
-			slog.Info(fmt.Sprintf("%s already installed", ext.Name))
+			slog.Info(fmt.Sprintf("Extension %s already installed", ext.Name))
 			return nil
 		}
 		if err := install(ctx, hcli, ext); err != nil {
-			return errors.Wrapf(err, "install %s", ext.Name)
+			return errors.Wrap(err, "install")
 		}
+		return nil
+	})
+}
 
-	case actionUpgrade:
+func handleExtensionUpgrade(ctx context.Context, kcli client.Client, hcli helm.Client, in *ecv1beta1.Installation, ext ecv1beta1.Chart, needsUpgrade bool) (finalErr error) {
+	return handleExtension(ctx, kcli, in, ext, actionUpgrade, func() error {
+		if !needsUpgrade {
+			slog.Info(fmt.Sprintf("%s already up to date", ext.Name))
+			return nil
+		}
 		if err := upgrade(ctx, hcli, ext); err != nil {
-			return errors.Wrapf(err, "upgrade %s", ext.Name)
+			return errors.Wrap(err, "upgrade")
 		}
+		return nil
+	})
+}
 
-	case actionUninstall:
+func handleExtensionUninstall(ctx context.Context, kcli client.Client, hcli helm.Client, in *ecv1beta1.Installation, ext ecv1beta1.Chart) (finalErr error) {
+	return handleExtension(ctx, kcli, in, ext, actionUninstall, func() error {
 		exists, err := hcli.ReleaseExists(ctx, ext.TargetNS, ext.Name)
 		if err != nil {
 			return errors.Wrap(err, "check if release exists")
 		}
 		if !exists {
-			slog.Info(fmt.Sprintf("%s already uninstalled", ext.Name))
+			slog.Info(fmt.Sprintf("Extension %s already uninstalled", ext.Name))
 			return nil
 		}
 		if err := uninstall(ctx, hcli, ext); err != nil {
-			return errors.Wrapf(err, "uninstall %s", ext.Name)
+			return errors.Wrap(err, "uninstall")
+		}
+		return nil
+	})
+}
+
+func handleExtension(ctx context.Context, kcli client.Client, in *ecv1beta1.Installation, ext ecv1beta1.Chart, action helmAction, processFn func() error) (finalErr error) {
+	processed, err := extensionAlreadyProcessed(ctx, kcli, in, ext)
+	if err != nil {
+		return errors.Wrap(err, "check if extension is already processed")
+	} else if processed {
+		slog.Info(fmt.Sprintf("Extension %s already processed!", ext.Name))
+		return nil
+	}
+
+	actionIng, _ := formatAction(action)
+
+	slog.Info(fmt.Sprintf("%s extension %s", actionIng, ext.Name))
```

Comment:
why not keep logging the version like before?

Reply:
I am just trying to make it consistent. Should we go with all structured logging?

Other instances where it is not structured:

```
			slog.Info(fmt.Sprintf("%s already installed", ext.Name))
```

```
		slog.Info(fmt.Sprintf("%s is ready!", ext.Name))
```

Reply:
we should go with structured logging, but slog doesn't provide wrapping out of the box. this is how i had it before:
```golang
slog.Info(fmt.Sprintf("%s extension", actionIng), "name", ext.Name, "version", ext.Version)
```